### ISSUE TITLE
testdouble causes bluebird to emit Unhandled rejection

### ISSUE THREAD

**Issue by queston02** [testdouble/testdouble.js#168]
When we use _td.matchers.anything()_ with _thenReject_ bluebird starts emiting "Unhandled rejection" message in console. Below is an example ...

td.when(someService.doSomething(td.matchers.anything())).thenReject('Some Error');

If in the above code i replace _td.matchers.anything()_ with exact parameter then the messages stops. I think testdouble is not able to notify bluebird that promise rejection is handled when we are using _td.matchers.anything()_  

Let me know if you need full code from me to reproduce this.

**Comment by searls** [comment]
A couple thoughts here:

1. I think going forward for edge case questions like this, adding a failing example to the `example/` directory (to an existing project if possible, new project if necessary), will really help us a lot to save on up-front communication about what the problem is. In this case, you should be able to add a failing example to the `example/node` project by installing bluebird there.

2. Yeah, what you're describing sounds broken, but I don't use bluebird a whole lot.

3. @schoonology, when you talked about a default issue template, I'd 100% support that if it also asked for an example test in the case of bugs/unexpected behavior.

**Comment by terinjokes** [comment]
I'm using the native Promises in recent versions of Node.js and get a similar warning.

```
(node:3358) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): undefined
(node:3358) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I'm not sure if this is a bad pattern, but I've recently been wiring things up with code similar to the following.

```javascript
td.when(getter.got(), {ignoreExtraArgs: true}).thenReject();
td.when(getter.got('path/to/something').thenResolve(body);
```

Otherwise, the test failures were pointing me in the wrong direction (complaining about the subject calling `.then` on undefined) when not implemented correctly.

**Comment by cjthompson** [comment]
I'm having an issue with `thenReject` also causing an unhandled rejection. 

The code is pretty straight forward. 

```
class Cache {
  function check() {
    return status.getDeployDate()
      .then(...)
      .catch((e) => this.emit('error'));
  }
}
```
and the test
```
  const status = td.replace('../../../src/lib/status');
  td.when(status.getDeployDate()).thenReject(new Error());
  const cache  = new Cache();
  cache.on('error', (e) => assert(e.message, 'error'));
  cache().then(() => done());
```

The test passes fine. The event handler fires and the assert works correctly.  Clearly in the Cache code, there is a `.catch`, so the Promise rejection is being caught.  I have no idea where the unhandled rejection is coming from.

**Comment by searls** [comment]
@cjthompson are you saying that if you call `td.reset()` between tests that you no longer have the issue?

**Comment by terinjokes** [comment]
@searls I call `td.reset()` in an `afterEach` and I do see the behavior described here. Would be curious to know if @cjthompson can reproduce this.

**Comment by cjthompson** [comment]
I figured that I didn't need to because each `td.when()` defines a return value, so as long as each test consumes the function call, the next test should be able to just call `td.when()` again without a reset.

I ended up having to do this because I'm using `ava` as a test runner and it runs all the tests concurrently by default, which doesn't work when you're mocking dependencies, so I serialized the tests and added a `td.reset()` in the `afterEach` and now they work fine.

I don't know why `td.reset()` would fix the issue unless there is a leak somewhere in the `testdouble` code that's leaking the rejected promise.  I didn't see anything obvious when I looked through the code though.

**Comment by searls** [comment]
Closing due to inactivity. If someone has a repeatable example I'll be glad to look at anew!

**Comment by pke** [comment]
I can still reproduce this @cjthompson and I am also using ava.

Given this code:

```es6
const test = require('ava');
const td = require('testdouble');
const LastCommitLog = td.replace('last-commit-log');
const parseAppInfo = require('..');

test('does not include missing git info', async t => {
  td.when(LastCommitLog.prototype.getLastCommit()).thenReject();
  const appInfo = await parseAppInfo();
  t.is(appInfo.hash, undefined);
  t.is(appInfo.tag, undefined);
});
```

The test will run successfully but the ava output will read:

```bash
√ test » does not include missing git info

  Unhandled rejection in test\test.js

  undefined


  1 test passed
  1 unhandled rejection
```

Its my understanding, that ava should not even see the rejection as its only executed when the mocked function is called.

Code available [here](https://github.com/pke/parse-app-info/tree/fix/proper-tests-and-coverage)

**Comment by searls** [comment]
[this link](
https://github.com/pke/parse-app-info/blob/fix/proper-tests-and-coverage/test/test.js) in your edit is 404'ing so I'm not 100% sure what you're referring to, but I imagine this is the problem:

```js
const LastCommitLog = td.replace('last-commit-log');
//..
test('does not include missing git info', async t => {
  //...
});
```

Because `td.replace` is designed to be cleaned up by `td.reset()` on an `afterEach` basis, it's not appropriate to call it from the top-level of a module, which are typically cached once `require`'d, and you'll wind up with all kinds of no-fun test pollution. ava's default behavior of splitting one-process-per-test-file confuses a lot of people on this point. The solution is to call `td.replace` and assign its return value to a variable from the test itself or from a `beforeEach` and to be sure you're globally calling `td.reset()` somewhere in your test config.

Relevant part of the [README](https://github.com/testdouble/testdouble.js#module-replacement-with-nodejs)

**Comment by pke** [comment]
Sorry the branch the link pointed to was merged and deleted already.
https://github.com/pke/parse-app-info/commit/6ee58c9cbec75e15e9b3aae8b03dab6cb5738965

Essentially I figured out what you analysed myself then. Moved everything into tests helped to solve the problem.
The README part I read and decided to go with a slightly cleaner way to hand down the imported files and subject to the tests (context instead of test file global vars).

So far I like testdouble, it was really painless to add after all.
